common_test.go

```diff
@@ -81,12 +81,81 @@ func assertCipherSuiteParamsEquals(t *testing.T, a, b CipherSuiteParams) {
 
 func assertDeepEquals(t *testing.T, a, b interface{}) {
 	t.Helper()
-	assert(t, reflect.DeepEqual(a, b), fmt.Sprintf("%+v != %+v", a, b))
+	assertTrue(t, reflect.DeepEqual(a, b), fmt.Sprintf("%+v != %+v", a, b))
 }
 
 func assertSameType(t *testing.T, a, b interface{}) {
 	t.Helper()
 	A := reflect.TypeOf(a)
 	B := reflect.TypeOf(b)
-	assert(t, A == B, fmt.Sprintf("%s != %s", A.Name(), B.Name()))
+	assertTrue(t, A == B, fmt.Sprintf("%s != %s", A.Name(), B.Name()))
+}
+
+// Utilities for parametrized tests
+// Represents the configuration for a given test instance.
+type testInstanceState interface {
+	set(string, string)
+	get(string) string
+}
+
+// This complicated mixin stuff is in case we want to extend this to
+// pass other parameters that aren't srings. Arguably YAGNI.
+type testInstanceStateMixin struct {
+	params map[string]string
+}
+
+func (m *testInstanceStateMixin) set(k string, v string) {
+	m.params[k] = v
+}
+
+func (m *testInstanceStateMixin) get(k string) string {
+	return m.params[k]
+}
+
+func newTestInstanceStateMixin() testInstanceStateMixin {
+	return testInstanceStateMixin{make(map[string]string, 0)}
+}
+
+type testInstanceStateBase struct {
+	testInstanceStateMixin
+}
+
+func newTestInstanceStateBase() testInstanceState {
+	return &testInstanceStateBase{newTestInstanceStateMixin()}
+}
+
+// Helper function.
+func runParametrizedInner(t *testing.T, name string, state testInstanceState, inparams []testParameter, f parametrizedTest) {
+	next := inparams[1:]
+
+	param := inparams[0]
+	for _, p := range param.vals {
+		state.set(param.name, p)
+		var n string
+		if len(name) > 0 {
+			n = name + "/"
+		}
+		n = n + param.name + "=" + p
+
+		if len(next) == 0 {
+			t.Run(n, func(t *testing.T) {
+				f(t, n, state)
+			})
+			continue
+		}
+		runParametrizedInner(t, n, state, next, f)
+	}
+}
+
+// Nominally public API.
+type testParameter struct {
+	name string
+	vals []string
+}
+
+type parametrizedTest func(t *testing.T, name string, p testInstanceState)
+
+// This is the function you call.
+func runParametrizedTest(t *testing.T, inparams []testParameter, f parametrizedTest) {
```

Comment:
I think I might need a little more explanation here.  Let me check my understanding here:

* `inparams` is basically a `map[string][]string`, where the key is the name of a parameter, and the array are its possible values
* For each parameter, you iterate through its values, running the test on a params struct with that value
* At the end of that loop, you leave that parameter on its last value, and move on to the next param

So given the following params dict:

```
{
  {"a", {"x", "y", "z"}},
  {"b", {"r", "s"}},
  {"c", {"q"}},
}
```

... you would go through the following cases

* a="x" b="" c=""
* a="y" b="" c=""
* a="z" b="" c=""
* a="z" b="" c=""
* a="z" b="r" c=""
* a="z" b="s" c=""
* a="z" b="s" c="q"

Assuming I've got all that right... That seems kind of odd, for two reasons.  First, you have the default values for things in the beginning, then you've got things stuck on their last value.

ISTM there are basically two sensible options here:
* Test all the combinations, or if you want to keep things linear
* Set a default value for each parameter, which is used outside of its test run

So with the above, you would have:

* a="x" b=default c=default
* a="y" b=default c=default
* a="z" b=default c=default
* a="z" b=default c=default
* a=default b="r" c=default
* a=default b="s" c=default
* a=default b=default c="q"

You could also treat "" as the default and leave it up to the test to deal with it.  In that case, the only real change you would need to do would be `state.set(param.name, "")` in `runParametrizedInner()`.

In any case, it would be good to have a brief explanation in a comment here about how this all works.

Reply:
Ah, no. This is supposed to run all the combinations for each parameter. So, given the following test:

```
func testParametrized(t *testing.T, n string, s testInstanceState) {
	fmt.Println(n, s)
}

func TestParametrized(t *testing.T) {
	p := []testParameter{
		testParameter{"a", []string{"a1", "a2"}},
		testParameter{"b", []string{"b1", "b2"}},
		testParameter{"c", []string{"c1", "c2"}},
	}

	runParametrizedTest(t, p, testParametrized)
}

````

You get the output:

```
a=a1/b=b1/c=c1 &{{map[c:c1 a:a1 b:b1]}}
a=a1/b=b1/c=c2 &{{map[c:c2 a:a1 b:b1]}}
a=a1/b=b2/c=c1 &{{map[a:a1 b:b2 c:c1]}}
a=a1/b=b2/c=c2 &{{map[a:a1 b:b2 c:c2]}}
a=a2/b=b1/c=c1 &{{map[b:b1 c:c1 a:a2]}}
a=a2/b=b1/c=c2 &{{map[c:c2 a:a2 b:b1]}}
a=a2/b=b2/c=c1 &{{map[a:a2 b:b2 c:c1]}}
a=a2/b=b2/c=c2 &{{map[a:a2 b:b2 c:c2]}}
```

Perhaps you're getting confused by len(next), but that's about the length of the remaining parameter list to iterate over.

Reply:
Yep, this is my error.  Thanks for the explanation.